firebase-ai-friendly-meals/android/app/src/main/java/com/google/firebase/example/friendlymeals/data/datasource/DatabaseRemoteDataSource.kt

```diff
@@ -0,0 +1,238 @@
+
+        saveRef.set(save).await()
+
+        firestore
+            .collection(RECIPE_COLLECTION)
+            .document(save.recipeId)
+            .update(SAVES_FIELD, FieldValue.increment(1))
+            .await()
+    }
+
+    suspend fun removeFavorite(save: Save) {
+        firestore
+            .collection(SAVE_COLLECTION)
+            .document("${save.recipeId}_${save.userId}")
+            .delete()
+            .await()
+
+        firestore
+            .collection(RECIPE_COLLECTION)
+            .document(save.recipeId)
+            .update(SAVES_FIELD, FieldValue.increment(-1))
+            .await()
+    }
+
+    suspend fun getFavorite(userId: String, recipeId: String): Boolean {
+        val document = firestore.collection(SAVE_COLLECTION)
+            .document("${recipeId}_${userId}")
+            .get()
+            .await()
+            .toObject<Save>()
+
+        return document != null
+    }
+
+    suspend fun getFilteredRecipes(
+        filterOptions: FilterOptions,
+        userId: String
+    ): List<RecipeListItem> {
+        var pipeline = firestore.pipeline().collection(RECIPE_COLLECTION)
+
+        if (filterOptions.recipeTitle.isNotBlank()) {
+            pipeline = pipeline
+                .where(field(TITLE_FIELD).toLower()
+                    .stringContains(filterOptions.recipeTitle))
+        }
+
+        if (filterOptions.filterByMine) {
+            pipeline = pipeline
+                .where(field(AUTHOR_ID_FIELD)
+                    .equal(userId))
+        }
+
+        if (filterOptions.rating > 0) {
+            pipeline = pipeline
+                .where(field(AVERAGE_RATING_FIELD)
+                    .greaterThanOrEqual(filterOptions.rating))
+        }
+
+        if (filterOptions.selectedTags.isNotEmpty()) {
+            pipeline = pipeline
+                .where(field(TAGS_FIELD)
+                    .arrayContainsAny(filterOptions.selectedTags))
+        }
+
+        when (filterOptions.sortBy) {
+            SortByFilter.RATING -> {
+                pipeline = pipeline
+                    .sort(field(AVERAGE_RATING_FIELD)
+                        .descending())
+            }
+            SortByFilter.ALPHABETICAL -> {
+                pipeline = pipeline
+                    .sort(field(TITLE_FIELD)
+                        .ascending())
+            }
+            SortByFilter.POPULARITY -> {
+                pipeline = pipeline
+                    .sort(field(SAVES_FIELD)
+                        .descending())
+            }
+        }
+
+        val results = pipeline.execute().await().results
+
+        return results.map {
+            val itemData = it.getData()
+
+            RecipeListItem(
+                id = itemData["id"] as? String ?: "",
+                title = itemData["title"] as? String ?: "",
+                averageRating = itemData["averageRating"] as? Double ?: 0.0,
+                imageUri = itemData["imageUri"] as? String
+            )
+        }
```

Comment:
nit: you can `filter` out any `RecipeListItem`s with empty IDs, and maybe log an error when a snapshot doesn't successfully map to a `RecipeListItem`.